<a href="https://colab.research.google.com/github/ibrahimalhanifsubhan-sketch/msci-151/blob/main/CW4_IAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Read the Excel file "/content/MSCI151 Data.xlsx" and print the header.

## Install necessary libraries

### Subtask:
Install the `openpyxl` library to read Excel files.


**Reasoning**:
Install the openpyxl library using pip.



In [ ]:
%pip install openpyxl

## Load the data

### Subtask:
Load the data from the provided Excel file path into a pandas DataFrame.


**Reasoning**:
Load the data from the Excel file into a pandas DataFrame.



In [ ]:
import pandas as pd

df = pd.read_excel("/content/MSCI151 Data.xlsx")

## Display the heading

### Subtask:
Print the header of the DataFrame.


**Reasoning**:
Display the header of the dataframe to show the column names.



In [ ]:
display(df.head())

id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  
0           9.16              2           0  
1          41.42              1           0  
2           5.88              3           1  
3          11.08              2           0  
4          24.61              1           0

In [ ]:
print(df.columns)

Index(['id', 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size',
       'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender'],
      dtype='object')


## Summary:

### Data Analysis Key Findings

*   The header of the DataFrame was successfully displayed, showing the column names of the loaded data.

### Insights or Next Steps

*   The data has been loaded and the header inspected, the next step would be to explore the data further, such as checking data types, missing values, and summary statistics, to understand its structure and content.


# Task
Analyze the dataset located at "/content/MSCI151 Data.xlsx". Perform predictive modeling using Random Forest Regressor. The target variables are 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender'. Use an 80/20 train-test split for model evaluation.

## Prepare the data

### Subtask:
Define the features (X) and the target variables (y) from the DataFrame.


**Reasoning**:
Define the list of target variables and then define the feature and target dataframes.



In [ ]:
target_variables = ['Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender']
X = df.drop(['id'] + target_variables, axis=1)
y = df[target_variables]

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Split the data into training and testing sets using an 80/20 ratio.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train and evaluate models

### Subtask:
Iterate through each target variable, train a Random Forest Regressor model for the current target variable, and evaluate the model on the test set using appropriate metrics (e.g., Mean Squared Error, R-squared).


**Reasoning**:
Iterate through each target variable, train a Random Forest Regressor model, and evaluate the model on the test set using MSE and R2 score, storing the results in a dictionary.



In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

results = {}

for target in target_variables:
    model = RandomForestRegressor(random_state=42)
    model.fit(X_train, y_train[target])
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test[target], y_pred)
    r2 = r2_score(y_test[target], y_pred)
    results[target] = {'MSE': mse, 'R2': r2}

display(results)

ValueError: at least one array or dtype is required

**Reasoning**:
The traceback indicates a ValueError in `sklearn.utils.validation.check_array`, suggesting an issue with the input data `X_train`. This is likely because `X_train` is an empty DataFrame, as seen in the provided variable information. The previous steps successfully split the data, but it appears the definition of X was incorrect, resulting in an empty DataFrame. Need to correct the definition of X to include the feature columns.



In [ ]:
X = df.drop(['id'] + target_variables, axis=1)
y = df[target_variables]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

results = {}

for target in target_variables:
    model = RandomForestRegressor(random_state=42)
    model.fit(X_train, y_train[target])
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test[target], y_pred)
    r2 = r2_score(y_test[target], y_pred)
    results[target] = {'MSE': mse, 'R2': r2}

display(results)

ValueError: at least one array or dtype is required

**Reasoning**:
The error persists, indicating that `X_train` is still an empty DataFrame. This means the definition of `X` is still incorrect. The features should be all columns except 'id' and the target variables. Let's redefine X and y and then proceed with the model training and evaluation.



In [ ]:
X = df.drop(['id'] + target_variables, axis=1)
y = df[target_variables]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

results = {}

for target in target_variables:
    model = RandomForestRegressor(random_state=42)
    model.fit(X_train, y_train[target])
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test[target], y_pred)
    r2 = r2_score(y_test[target], y_pred)
    results[target] = {'MSE': mse, 'R2': r2}

display(results)

ValueError: at least one array or dtype is required

**Reasoning**:
The error message "ValueError: at least one array or dtype is required" indicates that the input to `check_array` within the `fit` method is empty or invalid. Looking at the previous steps and the provided variable information, the `X` DataFrame is indeed empty. This is because the previous code attempted to drop all columns except 'id' and the target variables, leaving no features. To fix this, the definition of `X` should be all columns *except* 'id', and the target variables should be defined separately as `y`. The previous steps for splitting and modeling are correct once `X` and `y` are properly defined.



In [ ]:
X = df.drop('id', axis=1)
y = df[target_variables]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

results = {}

for target in target_variables:
    model = RandomForestRegressor(random_state=42)
    model.fit(X_train, y_train[target])
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test[target], y_pred)
    r2 = r2_score(y_test[target], y_pred)
    results[target] = {'MSE': mse, 'R2': r2}

display(results)

{'Profitability': {'MSE': 0.6614720716000017, 'R2': 0.9856346289663989},
 'debt': {'MSE': 1.8067554832000323, 'R2': 0.9936365019075044},
 'RnD': {'MSE': 3175.7946259452037, 'R2': 0.5453748931421671},
 'Firm Age': {'MSE': 0.0017078895999999974, 'R2': 0.9931575111345868},
 'Firm Size': {'MSE': 0.05764124240000063, 'R2': 0.9798572765746264},
 'CEO Experience': {'MSE': 0.004964000000000003, 'R2': 0.9891520979020979},
 'foreign sales': {'MSE': 0.5796764047999883, 'R2': 0.9992883852544107},
 'CEO education': {'MSE': 0.0, 'R2': 1.0},
 'CEO Gender': {'MSE': 0.0, 'R2': 1.0}}

## Report the results

### Subtask:
Print the evaluation metrics for each model.


**Reasoning**:
Iterate through the results dictionary and print the evaluation metrics for each target variable.



In [ ]:
for target, metrics in results.items():
    print(f"Target Variable: {target}")
    print(f"  Mean Squared Error (MSE): {metrics['MSE']:.4f}")
    print(f"  R-squared (R2): {metrics['R2']:.4f}")

Target Variable: Profitability
  Mean Squared Error (MSE): 0.6615
  R-squared (R2): 0.9856
Target Variable: debt
  Mean Squared Error (MSE): 1.8068
  R-squared (R2): 0.9936
Target Variable: RnD
  Mean Squared Error (MSE): 3175.7946
  R-squared (R2): 0.5454
Target Variable: Firm Age
  Mean Squared Error (MSE): 0.0017
  R-squared (R2): 0.9932
Target Variable: Firm Size
  Mean Squared Error (MSE): 0.0576
  R-squared (R2): 0.9799
Target Variable: CEO Experience
  Mean Squared Error (MSE): 0.0050
  R-squared (R2): 0.9892
Target Variable: foreign sales
  Mean Squared Error (MSE): 0.5797
  R-squared (R2): 0.9993
Target Variable: CEO education
  Mean Squared Error (MSE): 0.0000
  R-squared (R2): 1.0000
Target Variable: CEO Gender
  Mean Squared Error (MSE): 0.0000
  R-squared (R2): 1.0000


## Summary:

### Data Analysis Key Findings

*   Random Forest Regressor models were trained and evaluated for nine different target variables: 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender'.
*   The R-squared scores for the trained models varied significantly across the target variables, with some models demonstrating very high predictive capability (R2 close to 1.0) and others having lower R2 scores.
*   The Mean Squared Error (MSE) also varied for each target variable, reflecting the average squared difference between the actual and predicted values.

### Insights or Next Steps

*   Further analysis of the feature importance from each Random Forest model could provide insights into which features are most influential in predicting each target variable.
*   Consider exploring other regression models or hyperparameter tuning for the Random Forest Regressor to potentially improve the prediction performance, especially for target variables with lower R2 scores.


# Task
Read the data from the file "/content/New.xlsx" and print the heading.

## Install necessary libraries

### Subtask:
Ensure the `openpyxl` library is installed to read Excel files.


**Reasoning**:
Install the openpyxl library using pip.



In [ ]:
%pip install openpyxl

## Load the data

### Subtask:
Load the data from the provided Excel file path into a pandas DataFrame.


**Reasoning**:
Read the Excel file into a pandas DataFrame.



In [ ]:
df = pd.read_excel("/content/New.xlsx")

## Display the heading

### Subtask:
Print the header of the DataFrame.


**Reasoning**:
Print the first few rows of the DataFrame `df` to display the header and some sample data, which fulfills the subtask requirement of printing the header.



In [ ]:
display(df.head())

id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  Marketing  Produnt_Sentimet  \
0           9.16              2           0   0.119645          0.674037   
1          41.42              1           0   0.468057          0.517517   
2           5.88              3           1   0.894631          0.261792   
3          11.08              2           0   0.815249          0.241329   
4          24.61              1           0   0.948117          0.705435   

   Politiciants  Unnamed: 13  
0      0.030721            0  
1      0.228008            0  
2      0.429228            0  
3      0.267118            0  
4      0.409645            0

## Summary:

### Data Analysis Key Findings

* The `openpyxl` library, necessary for reading Excel files, was already installed.
* The data from the "/content/New.xlsx" file was successfully loaded into a pandas DataFrame named `df`.
* The header and the first five rows of the DataFrame were displayed.

### Insights or Next Steps

* The data is ready for further analysis or manipulation.
* The next step could involve exploring the data types of the columns or getting a summary of the data.


# Task
Build a predictive model using the Decision Tree algorithm to predict 'Profitability' (Y) based on the following features (X): 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender', 'marketing', 'product sentiment', 'POB categorical'. Split the data into training and testing sets using an 80/20 ratio using data from "/content/New.xlsx".

## Prepare the data

### Subtask:
Define the features (X) and the target variable (y) from the DataFrame.


**Reasoning**:
Define the feature and target columns and create the corresponding dataframes.



In [ ]:
feature_columns = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Produnt_Sentimet', 'Politiciants']
target_column = 'Profitability'
X = df[feature_columns]
y = df[target_column]

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Split the data into training and testing sets using an 80/20 ratio with a fixed random state for reproducibility.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train the model

### Subtask:
Train a Decision Tree Regressor model on the training data.


**Reasoning**:
Import the necessary class and train the Decision Tree Regressor model on the training data.



In [ ]:
from sklearn.tree import DecisionTreeRegressor

model = DecisionTreeRegressor(random_state=42)
model.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

## Evaluate the model

### Subtask:
Evaluate the trained model on the test set using appropriate metrics (e.g., Mean Squared Error, R-squared).


**Reasoning**:
Evaluate the trained Decision Tree Regressor model on the test set by calculating the Mean Squared Error (MSE) and R-squared (R2) score.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"R-squared (R2): {r2:.4f}")

Mean Squared Error (MSE): 107.0053
R-squared (R2): -1.3239


## Report the results

### Subtask:
Print the evaluation metrics.


**Reasoning**:
Print the Mean Squared Error (MSE) and R-squared (R2) values that were calculated and stored in the `mse` and `r2` variables in the previous step.



In [ ]:
print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"R-squared (R2): {r2:.4f}")

Mean Squared Error (MSE): 107.0053
R-squared (R2): -1.3239


## Summary:

### Data Analysis Key Findings

*   The Mean Squared Error (MSE) of the Decision Tree Regressor model on the test set is approximately 107.0053.
*   The R-squared (\text{R}^2) score of the model on the test set is approximately -1.3239.

### Insights or Next Steps

*   The negative \text{R}^2 score indicates that the current Decision Tree Regressor model performs poorly and is worse than simply predicting the mean of the target variable. Further model tuning or selection of a different algorithm is necessary.
*   Investigate feature importance from the trained Decision Tree model to understand which features contribute most to the predictions and potentially inform feature selection or engineering for improved performance.


## Feature Importance

### Subtask:
Display the feature importance of the Decision Tree Regressor model.

**Reasoning**:
Access the `feature_importances_` attribute of the trained Decision Tree Regressor model and display the feature importance scores.

In [ ]:
importances = model.feature_importances_

# Create a dictionary to store feature names and their importance scores
feature_importance_dict = dict(zip(X.columns, importances))

# Sort the features by importance
sorted_feature_importance = sorted(feature_importance_dict.items(), key=lambda item: item[1], reverse=True)

# Print the feature importance
print("Feature Importance (Decision Tree Regressor):")
for feature, importance in sorted_feature_importance:
    print(f"{feature}: {importance:.4f}")

Feature Importance (Decision Tree Regressor):
debt: 0.3061
RnD: 0.2339
CEO Gender: 0.1597
Politiciants: 0.1096
Marketing: 0.0862
CEO Experience: 0.0437
Produnt_Sentimet: 0.0195
Firm Size: 0.0189
Firm Age: 0.0112
foreign sales: 0.0098
CEO education: 0.0012


## Feature Selection using Lasso

### Subtask:
Apply Lasso regularization to the training data to identify important features and display the selected features.

**Reasoning**:
Import the Lasso model, fit it to the training data, and then identify and display the features with non-zero coefficients as these are selected by Lasso regularization.

In [ ]:
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler

# Scale the data before applying Lasso
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply Lasso with cross-validation to find the best alpha
lasso = LassoCV(cv=5, random_state=42)
lasso.fit(X_train_scaled, y_train)

# Get the coefficients
lasso_coef = lasso.coef_

# Identify selected features (coefficients that are not zero)
selected_features = X.columns[(lasso_coef != 0)]

print("Features selected by Lasso regularization:")
print(selected_features)
print(lasso_coef)

Features selected by Lasso regularization:
Index(['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience',
       'foreign sales', 'CEO education', 'CEO Gender', 'Marketing',
       'Produnt_Sentimet'],
      dtype='object')
[-3.6299501   2.12724543 -0.79611312  0.14136795  1.26416836 -0.16860013
  1.34429396  4.53516079  1.73420029 -0.96927459 -0.        ]


# Task
Read the data from the Excel file "/content/New_.xlsx" and print the heading.

## Install necessary libraries

### Subtask:
Ensure the `openpyxl` library is installed to read Excel files.


**Reasoning**:
Install the openpyxl library using pip.



In [ ]:
%pip install openpyxl

## Load the data

### Subtask:
Load the data from the provided Excel file path into a pandas DataFrame.


**Reasoning**:
Read the Excel file into a pandas DataFrame.



In [ ]:
df = pd.read_excel("/content/New_.xlsx")

## Display the heading

### Subtask:
Print the header of the DataFrame.


**Reasoning**:
Print the first few rows of the DataFrame `df` to display the header and some sample data, which fulfills the subtask requirement of printing the header.



In [ ]:
display(df.head())

id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  Marketing  Produnt_Sentimet  \
0           9.16              2           0   0.119645          0.674037   
1          41.42              1           0   0.468057          0.517517   
2           5.88              3           1   0.894631          0.261792   
3          11.08              2           0   0.815249          0.241329   
4          24.61              1           0   0.948117          0.705435   

   Politiciants  PoB  
0      0.030721    0  
1      0.228008    0  
2      0.429228    0  
3      0.267118    0  
4      0.409645    0

## Summary:

### Data Analysis Key Findings

*   The data from the Excel file "/content/New\_.xlsx" was successfully loaded into a pandas DataFrame.
*   The header of the DataFrame was successfully displayed using `df.head()`.

### Insights or Next Steps

*   The next logical step is to explore the data further by looking at its shape, data types, and summary statistics.
*   Based on the content of the header, the subsequent analysis can focus on specific columns of interest.


# Task
Build a predictive model using the Decision Tree algorithm to predict 'Profitability' (Y) based on the following features (X): 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender', 'marketing', 'product sentiment', 'POB categorical'. Split the data into training and testing sets using an 80/20 ratio.

## Prepare the data

### Subtask:
Define the features (X) and the target variable (y) from the DataFrame.


**Reasoning**:
Define the feature and target columns and create the corresponding dataframes.



In [ ]:
feature_columns = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Produnt_Sentimet', 'PoB']
target_column = 'Profitability'
X = df[feature_columns]
y = df[target_column]

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Split the data into training and testing sets using an 80/20 ratio with a fixed random state for reproducibility.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train the model

### Subtask:
Train a Decision Tree Regressor model on the training data.


**Reasoning**:
Train the Decision Tree Regressor model on the training data.



In [ ]:
from sklearn.tree import DecisionTreeRegressor

model = DecisionTreeRegressor(random_state=42)
model.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

## Evaluate the model

### Subtask:
Evaluate the trained model on the test set using appropriate metrics (e.g., Mean Squared Error, R-squared).


**Reasoning**:
Evaluate the trained Decision Tree Regressor model on the test set by calculating the Mean Squared Error (MSE) and R-squared (R2) score, and then print these metrics, which addresses all steps of the current subtask.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"R-squared (R2): {r2:.4f}")

Mean Squared Error (MSE): 686.0945
R-squared (R2): -13.9001


## Report the results

### Subtask:
Print the evaluation metrics.


**Reasoning**:
Print the calculated MSE and R2 values.



In [ ]:
print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"R-squared (R2): {r2:.4f}")

Mean Squared Error (MSE): 686.0945
R-squared (R2): -13.9001


## Summary:

### Data Analysis Key Findings

*   The Mean Squared Error (MSE) of the Decision Tree model on the test set is 686.0945.
*   The R-squared (R2) score on the test set is -13.9001.

### Insights or Next Steps

*   A negative R-squared score indicates that the model is a very poor fit for the data and performs worse than a simple horizontal line through the mean of the response variable.
*   Further investigation is needed to understand why the model is performing so poorly, potentially involving feature engineering, trying different algorithms, or hyperparameter tuning.


## Feature Selection using Lasso

### Subtask:
Apply Lasso regularization to the training data to identify important features and display the selected features.

**Reasoning**:
Import the Lasso model, fit it to the training data, and then identify and display the features with non-zero coefficients as these are selected by Lasso regularization.

In [ ]:
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler

# Scale the data before applying Lasso
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply Lasso with cross-validation to find the best alpha
lasso = LassoCV(cv=5, random_state=42)
lasso.fit(X_train_scaled, y_train)

# Get the coefficients
lasso_coef = lasso.coef_

# Identify selected features (coefficients that are not zero)
selected_features = X.columns[(lasso_coef != 0)]

print("Features selected by Lasso regularization:")
print(selected_features)

Features selected by Lasso regularization:
Index(['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience',
       'foreign sales', 'CEO education', 'CEO Gender', 'Marketing',
       'Produnt_Sentimet', 'PoB'],
      dtype='object')


## Feature Importance

### Subtask:
Display the feature importance of the Decision Tree Regressor model.

**Reasoning**:
Access the `feature_importances_` attribute of the trained Decision Tree Regressor model and display the feature importance scores.

In [ ]:
importances = model.feature_importances_

# Create a dictionary to store feature names and their importance scores
feature_importance_dict = dict(zip(X.columns, importances))

# Sort the features by importance
sorted_feature_importance = sorted(feature_importance_dict.items(), key=lambda item: item[1], reverse=True)

# Print the feature importance
print("Feature Importance (Decision Tree Regressor):")
for feature, importance in sorted_feature_importance:
    print(f"{feature}: {importance:.4f}")

Feature Importance (Decision Tree Regressor):
debt: 0.3137
CEO education: 0.1869
CEO Gender: 0.1597
Firm Size: 0.1061
Marketing: 0.0823
CEO Experience: 0.0452
Firm Age: 0.0435
RnD: 0.0323
foreign sales: 0.0167
Produnt_Sentimet: 0.0104
PoB: 0.0031


In [38]:
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler

# Scale the data before applying Lasso
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply Lasso with cross-validation to find the best alpha
lasso = LassoCV(cv=5, random_state=42)
lasso.fit(X_train_scaled, y_train)

# Get the coefficients
lasso_coef = lasso.coef_

# Identify selected features (coefficients that are not zero)
selected_features = X.columns[(lasso_coef != 0)]

print("Features selected by Lasso regularization:")
print(selected_features)

Features selected by Lasso regularization:
Index(['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience',
       'foreign sales', 'CEO education', 'CEO Gender', 'Marketing',
       'Produnt_Sentimet', 'PoB'],
      dtype='object')


In [39]:
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler

# Scale the data before applying Lasso
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply Lasso with cross-validation to find the best alpha
lasso = LassoCV(cv=5, random_state=42)
lasso.fit(X_train_scaled, y_train)

# Get the coefficients
lasso_coef = lasso.coef_

# Identify selected features (coefficients that are not zero)
selected_features = X.columns[(lasso_coef != 0)]

print("Features selected by Lasso regularization:")
print(selected_features)

Features selected by Lasso regularization:
Index(['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience',
       'foreign sales', 'CEO education', 'CEO Gender', 'Marketing',
       'Produnt_Sentimet', 'PoB'],
      dtype='object')


In [42]:
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler

# Scale the data before applying Lasso
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply Lasso with cross-validation to find the best alpha
lasso = LassoCV(cv=5, random_state=42)
lasso.fit(X_train_scaled, y_train)

# Get the coefficients
lasso_coef = lasso.coef_

# Identify selected features (coefficients that are not zero)
selected_features = X.columns[(lasso_coef != 0)]

print("Features selected by Lasso regularization:")
print(selected_features)
display(lasso_coef)

Features selected by Lasso regularization:
Index(['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience',
       'foreign sales', 'CEO education', 'CEO Gender', 'Marketing',
       'Produnt_Sentimet', 'PoB'],
      dtype='object')


array([-3.61570998,  2.18324704, -0.86278446,  0.0961268 ,  1.32191929,
       -0.25683643,  1.35959146,  4.65859362,  1.79308102, -1.09797289,
       -0.52083171])

In [43]:
import statsmodels.api as sm

# Define the features (X) and the target variable (y)
feature_columns = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Produnt_Sentimet', 'PoB']
target_column = 'Profitability'
X = df[feature_columns]
y = df[target_column]

# Add a constant (intercept) to the feature set
X = sm.add_constant(X)

# Create and fit the OLS model with robust standard errors
model = sm.OLS(y, X)
results = model.fit(cov_type='HC1')  # HC1 is a common type of robust standard error

# Print the regression summary
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:          Profitability   R-squared:                       0.297
Model:                            OLS   Adj. R-squared:                  0.227
Method:                 Least Squares   F-statistic:                     5.014
Date:                Wed, 08 Oct 2025   Prob (F-statistic):           2.69e-06
Time:                        05:41:26   Log-Likelihood:                -463.99
No. Observations:                 122   AIC:                             952.0
Df Residuals:                     110   BIC:                             985.6
Df Model:                          11                                         
Covariance Type:                  HC1                                         
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const                1.1232     20.158  

In [45]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split

# Define the features (X) and the target variable (y)
feature_columns_subset = ['debt', 'CEO Gender']
X_subset = df[feature_columns_subset]
y = df[target_column] # Using the target variable defined previously

# Split the data into training and testing sets (although not strictly needed for feature importance, it's good practice)
X_train_subset, X_test_subset, y_train, y_test = train_test_split(X_subset, y, test_size=0.2, random_state=42)

# Train a Decision Tree Regressor model
model_subset = DecisionTreeRegressor(random_state=42)
model_subset.fit(X_train_subset, y_train)

# Get feature importances
importances_subset = model_subset.feature_importances_

# Create a dictionary to store feature names and their importance scores
feature_importance_dict_subset = dict(zip(X_subset.columns, importances_subset))

# Sort the features by importance
sorted_feature_importance_subset = sorted(feature_importance_dict_subset.items(), key=lambda item: item[1], reverse=True)

# Print the feature importance
print("Feature Importance (Decision Tree Regressor with selected features):")
for feature, importance in sorted_feature_importance_subset:
    print(f"{feature}: {importance:.4f}")

Feature Importance (Decision Tree Regressor with selected features):
debt: 0.8344
CEO Gender: 0.1656


In [46]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import pandas as pd # Import pandas here if it's not already imported

# Define the features for PCA
pca_features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Produnt_Sentimet', 'PoB']
X_pca = df[pca_features]

# Scale the features
scaler = StandardScaler()
X_pca_scaled = scaler.fit_transform(X_pca)

# Perform PCA
pca = PCA()
pca.fit(X_pca_scaled)

# Get the loading factors
loading_factors = pca.components_.T

# Create a DataFrame for better visualization of loading factors
loading_factors_df = pd.DataFrame(loading_factors, index=pca_features, columns=[f'PC{i+1}' for i in range(pca.n_components_)])

print("Loading Factors (Principal Component Analysis):")
display(loading_factors_df)

Loading Factors (Principal Component Analysis):


PC1       PC2       PC3       PC4       PC5       PC6  \
debt             -0.190357 -0.031849 -0.504949  0.006882 -0.151342  0.706752   
RnD               0.227566  0.212053 -0.288766  0.093347  0.565444 -0.073262   
Firm Age          0.360568  0.371086 -0.112079 -0.096731 -0.342632 -0.068861   
Firm Size         0.326622  0.197935  0.067200 -0.235748 -0.428472  0.147775   
CEO Experience    0.166816  0.536650  0.264807 -0.333729  0.156969  0.125136   
foreign sales    -0.129035 -0.015042  0.634236  0.106732  0.076805  0.571459   
CEO education     0.541395 -0.235416 -0.226202  0.301439 -0.036472  0.229236   
CEO Gender        0.560522 -0.262560  0.207889  0.025137  0.181826  0.129472   
Marketing        -0.117550  0.447042 -0.233149 -0.031194  0.306172  0.215115   
Produnt_Sentimet  0.040592  0.262984  0.151629  0.664754  0.140190 -0.001433   
PoB               0.089721 -0.309114 -0.021209 -0.518815  0.421971  0.103645   

                       PC7       PC8       PC9      PC10      PC11  
debt              0.094982 -0.197536  0.187774  0.181899 -0.262560  
RnD               0.607221 -0.090685  0.025665 -0.321272 -0.072375  
Firm Age         -0.330886 -0.486164  0.004854 -0.467144 -0.155373  
Firm Size         0.302809  0.648951  0.233711 -0.134271 -0.056051  
CEO Experience    0.079525 -0.261153  0.111781  0.564355  0.238467  
foreign sales     0.114137 -0.119025 -0.135042 -0.422633  0.113856  
CEO education    -0.097642  0.002961 -0.088179  0.061622  0.659122  
CEO Gender       -0.096913  0.024109 -0.281982  0.258614 -0.605670  
Marketing        -0.436367  0.437669 -0.448964 -0.085181  0.013894  
Produnt_Sentimet -0.255950  0.129581  0.579122  0.066051 -0.150332  
PoB              -0.356200  0.062930  0.504446 -0.219163  0.068122